### PR TITLE
#603, #446. Enhancing :- to allow multiple LHSs and return values

### LINKED ISSUE #446: :- syntax awkward for methods with multiple return values or no results
The error handling support in Dafny is currently centered around the "let or fail" statement, which attempts to `Extract()` a value and immediately returns the result of `PropagateFailure()` otherwise:

```
var x :- SomeMethod(...);
```

This doesn't compose well with methods that return multiple values, however, since `Extract()` must produce a single value, so you can never write this:

```
var x, y, z :- SomeMethod(...);
```

There is also use case for methods that operate by side-effects and return no values but may fail. I'm finding that I'm defining several methods that return a `Result<()>` of some kind and calling them like this:

```
var _ :- SomeMethod(...);
```

It's a minor annoyance but it adds up, especially when trying to implement tests. I'd suggest we support the unary variant of `:-` (currently only usable in functional expressions) for statements like this too:

```
:- SomeMethod(...);
```

### LINKED ISSUE #603: :- with more than one out-parameter
@samuelgruetter 

## Problem

If `M` is a method like

    method M(a: A) returns (res: Result<B>)

and `Result` is a type like:

    datatype Result<T> = Success(T) | Failure(string)
    {
      // members IsFailure, PropagateFailure, Extract
    }

then Dafny allows the method to be called like:

    b :- M(a);

There are examples where it would be useful to be able to return additional values. In particular, we have examples where it would be convenient to return ghost values together with the success value. It's awkward to do that today. For example, use of `:-` requires the method to have exactly one out-parameter, the tuple type is built in and does not have any form like

    (B, ghost C)

Therefore, one must declare a new type of pairs, like

    datatype ExtendWithGhost<X, Y> = ExtendWithGhost(x: X, ghost y: Y)

and use that in the signature of `M`:

    method M(a: A) returns (res: Result<ExtendWithGhost<B, C>>)

Not only is this construction a bit clumsy, but it also leaves an `ExtendWithGhost` wrapper around the `B` value at run time.

## Proposed solution

I propose that Dafny allow `:-` to be used to call methods with more than one out-parameter, provided that the first out-parameter follows the rules of the current `:-` method calls. (That is, the first out-parameter must be such a `Result` type.)

Here's what the example above would look like:

    method M(a: A) returns (res: Result<B>, ghost c: C)

A call would look like

    b, c :- M(a);

which would mean

    var tmp;
    tmp, c := M(a);
    if tmp.IsFailure() {
      callerRes := tmp.PropagateFailure();
      return;
    }
    b := tmp.Extract();

where `callerRes` is the name of the first out-parameter of the caller.

Note that the additional out-parameters (here, `c`) are passed regardless of the success/failure status of the first out-parameter.

Note that the way I defined the meaning above allows the caller itself to have any number of out-parameters, as long as the first one (which I assumed is called `callerRes`) has a type returned by `tmp.PropagateFailure()`. In particular, this means that `c` in the caller can be any l-value (for example, a local variable or any out-parameter of the caller).

In my example, I added one ghost parameter, which is a ghost, because this is scenario we have seen. Dafny could allow any number of additional out-parameters, both ghost and non-ghost.

## Alternative solution to the motivating example

Dafny could provide built-in support for tuples with ghost components. Such a type might be written as

    TupleType ::= "(" [ TupleComponent ( "," TupleComponent )+ ] ")"
    TupleComponent ::= [ "ghost" ] TypeName

Note that `TupleType` has 0 or 2-or-more components, just like tuples do today.

This new tuple type would compile to a tuple with all ghost components deleted. In the special case that there is only one non-ghost component, the type would compile to that component alone, without any 1-tuple wrapper.

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #603. Fixes #446.

Implements multiple return values and multiple LHSs in :- statements for both the cases of an Extract method and no Extract method.

Aspects still to do:
859 -- ghost checks
817 -- evaluate expressions in LHS before calling RHS method
817 -- l-values need to be distinct in the expression-RHS form, unless the RHS expressions are equal

### CODE REVIEW

**Comment by robin-aws** [comment]
Just flagging that I'd really like to take a closer look at this one before it's merged - I'm warming somewhat to the feature but not quite there yet :)

File: Source/Dafny/Resolver.cs [hunk 1]
```diff
@@ -11366,10 +11369,27 @@ private class RBranchExpr : RBranch {
 
       var temp = FreshTempVarName("valueOrError", codeContext);
       var tempType = new InferredTypeProxy();
-      s.ResolvedStatements.Add(
-        // "var temp := MethodOrExpression;"
-        new VarDeclStmt(s.Tok, s.Tok, new List<LocalVariable>() { new LocalVariable(s.Tok, s.Tok, temp, tempType, false) },
-          new UpdateStmt(s.Tok, s.Tok, new List<Expression>() { new IdentifierExpr(s.Tok, temp) }, new List<AssignmentRhs>() { new ExprRhs(s.Rhs) })));
+      var lhss = new List<LocalVariable>() { new LocalVariable(s.Tok, s.Tok, temp, tempType, false) }
+      ;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Resolver.cs]: Done

File: Source/Dafny/Resolver.cs [hunk 2]
```diff
@@ -11366,10 +11369,27 @@ private class RBranchExpr : RBranch {
...
+      if (s.Lhss.Count <= 1) {
+        s.ResolvedStatements.Add(
+          // "var temp := MethodOrExpression;"
+          new VarDeclStmt(s.Tok, s.Tok, lhss,
+            new UpdateStmt(s.Tok, s.Tok, new List<Expression>() {new IdentifierExpr(s.Tok, temp)},
+              new List<AssignmentRhs>() {new ExprRhs(s.Rhs)})));
+      } else {
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Resolver.cs]: Done

File: Source/Dafny/Resolver.cs [hunk 3]
```diff
@@ -11381,20 +11401,25 @@ private class RBranchExpr : RBranch {
...
-      if (s.Lhss.Count == 1)
+      if (s.Lhss.Count >= 1)
       {
         // "y := temp.Extract();"
         s.ResolvedStatements.Add(
-          new UpdateStmt(s.Tok, s.Tok, s.Lhss, new List<AssignmentRhs>() {
-            new ExprRhs(VarDotMethod(s.Tok, temp, "Extract"))}));
+          new UpdateStmt(s.Tok, s.Tok, 
+            new List<Expression>(){s.Lhss[0]}, 
+            new List<AssignmentRhs>(){new ExprRhs(VarDotMethod(s.Tok, temp, "Extract"))}
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Resolver.cs:11492]: OK

File: Source/Dafny/Resolver.cs [hunk 4]
```diff
@@ -11381,20 +11401,25 @@ private class RBranchExpr : RBranch {
           new IfStmt(s.Tok, s.Tok, false, VarDotMethod(s.Tok, temp, "IsFailure"),
             // THEN: { return temp.PropagateFailure(); }
             new BlockStmt(s.Tok, s.Tok, new List<Statement>() {
-              new ReturnStmt(s.Tok, s.Tok, new List<AssignmentRhs>() { new ExprRhs(VarDotMethod(s.Tok, temp, "PropagateFailure"))}),
+              new UpdateStmt(s.Tok, s.Tok, 
+                new List<Expression>() {new IdentifierExpr(s.Tok, (codeContext as Method).Outs[0].CompileName)},
+                new List<AssignmentRhs>() {new ExprRhs(VarDotMethod(s.Tok, temp, "PropagateFailure"))}
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Resolver.cs:11478]: OK

File: Test/git-issues/git-issue-446a.dfy [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  return Success(0);
+}
+
+
+method m(i: int) returns (r: Result<int>, o: int)
+  ensures i >= 0 ==> (r.Success? && r.value == i && o == i+i+i);
+  ensures i < 0 ==> (r.Failure? && o == i+i);
+{
+   if i < 0 { return Failure("negative"), i+i; }
+   return Success(i), i+i+i;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446a.dfy]: OK

File: Test/git-issues/git-issue-446a.dfy [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+   return Success(i);
+}
+
+method Main() {
+   var x := mn1();
+   print x.Failure?, " ";
+  // var out;
+  // x, out := mn();
+  // print x.Failure?, " ", out, " ";
+   print "End\n";
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446a.dfy]: OK

File: Test/git-issues/git-issue-446e.dfy [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+datatype Result<T> = Success(value: T) | Failure(error: string)
+{
+    predicate method IsFailure() {
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446e.dfy:6]: OK

File: Test/git-issues/git-issue-446a.dfy [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+   if i < 0 { return Failure("negative"), i+i; }
+   return Success(i), i+i+i;
+}
+
+method m1(i: int) returns (r: Result<int>)
+  ensures i >= 0 ==> (r.Success? && r.value == i);
+  ensures i < 0 ==> r.Failure? ;
+{
+   if i < 0 { return Failure("negative"); }
+   return Success(i);
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446a.dfy]: OK

File: Test/git-issues/git-issue-446a.dfy [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  out :- m(-1); // Should exit with failure
+  return Success(0);
+}
+*/
+method mn1() returns (r: Result<int>)
+  ensures r.Failure?
+{
+   :- m1(1);
+  print "OK\n";
+   :- m1(-1); // SHould exit with failure
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446a.dfy]: OK

File: Source/Dafny/Resolver.cs [hunk 10]
```diff
@@ -11381,20 +11401,25 @@ private class RBranchExpr : RBranch {
...
+              new ReturnStmt(s.Tok, s.Tok, null),
             }),
             // ELSE: no else block
             null
           ));
       }
 
-      Contract.Assert(s.Lhss.Count <= 1);
-      if (s.Lhss.Count == 1)
+      if (s.Lhss.Count >= 1)
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Resolver.cs]: OK

File: Source/Dafny/Resolver.cs [hunk 11]
```diff
@@ -11366,10 +11369,27 @@ private class RBranchExpr : RBranch {
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Resolver.cs:11390]: OK

File: Source/Dafny/Resolver.cs [hunk 12]
```diff
@@ -11381,20 +11401,25 @@ private class RBranchExpr : RBranch {
           new IfStmt(s.Tok, s.Tok, false, VarDotMethod(s.Tok, temp, "IsFailure"),
             // THEN: { return temp.PropagateFailure(); }
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Resolver.cs:11507]: Done

File: Test/exceptions/ParseErrors.dfy [hunk 13]
```diff
@@ -5,7 +5,7 @@ include "./VoidOutcome.dfy"
 
 method MultiAssignmentNotAllowed(r1: NatOutcome, r2: NatOutcome) returns (res: NatOutcome) {
     var a0, b0 := r1, r2; // <-- multi-assignment allowed
-    var a0, b0 :- r1, r2; // <-- multi-assignment not allowed
+    var a0, b0 :- r1, r2; // <-- multi-assignment now allowed
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/exceptions/ParseErrors.dfy]: Done

File: Test/git-issues/git-issue-446.dfy [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  var k :- m1(1);
+  assert k == 1;
+  out :- m1(-1); // Should exit with failure
+  return Success(k), out;
+}
+
+
+method m(i: int) returns (r: Result<int>, o: int)
+  ensures i >= 0 ==> (r.Success? && r.Extract() == i && o == i+i);
+  ensures i < 0 ==> (r.Failure? && o == i+i);
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446.dfy:48]: Done

File: Test/git-issues/git-issue-446.dfy [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+{
+   if i < 0 { return Failure("negative"), i+i; }
+   return Success(i), i+i;
+}
+
+method m1(i: int) returns (r: Result<int>)
+  ensures i >= 0 ==> (r.Success? && r.Extract() == i);
+  ensures i < 0 ==> r.Failure? ;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446.dfy:56]: Here and elsewhere

File: Test/git-issues/git-issue-446.dfy [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    function method PropagateFailure<U>(): Result<U>
+      requires Failure?
+    {
+      Failure(this.error)
+    }
+    function method Extract(): T
+      requires Success?
+    {
+      value
+    }
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446.dfy]: Here and elsewhere

File: Test/git-issues/git-issue-446a.dfy [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+datatype Result<T> = Success(value: T) | Failure(error: string)
+{
+    predicate method IsFailure() {
+      Failure?
+    }
+    function method PropagateFailure<U>(): Result<U>
+      requires Failure?
+    {
+      Failure(this.error)
+    }
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-446a.dfy]: Here and elsewhere

File: Test/git-issues/git-issue-505.dfy [hunk 18]
```diff
@@ -7,23 +7,23 @@ lemma a()
   var testbv: bv64 := a as bv64;
   var testval: int := testbv as int;
 
-  assert testval == a; // "Timed out on: assertion violation"
+  // assert testval == a; // "Timed out on: assertion violation"
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-505.dfy:10]: For the changes in this file, see also https://github.com/dafny-lang/dafny/pull/868/commits/e62b6e23d82d8eefe0799dfaba5b46e7fb8a88d1, which also had to adjust this test. Perhaps this test file should be removed altogether?

Comment (davidcok) [reply Test/git-issues/git-issue-505.dfy]: Fixed -- just including the small BV tests, which do not timeout

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**